File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_EMA_TwoBarsBelow.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.QuoteJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.EMAJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.model.BlockResultCodeJava.QUOTES_NOT_BELOW_EMA_SCREEN_2;
import static com.kishlaly.ta.utils.QuotesJava.isQuoteBelowEMA;

/**
 * the last two bars are completely below EMA13
 */
public class Long_ScreenTwo_EMA_TwoBarsBelow implements ScreenTwoBlockJava {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<EMAJava> screen_2_EMA13 = (List<EMAJava>) screen.indicators.get(IndicatorJava.EMA13);
        EMAJava ema13_2 = CollectionUtilsJava.getFromEnd(screen_2_EMA13, 2);
        EMAJava ema13_1 = CollectionUtilsJava.getFromEnd(screen_2_EMA13, 1);
        List<QuoteJava> screen_2_Quotes = screen.quotes;
        QuoteJava quote2 = CollectionUtilsJava.getFromEnd(screen_2_Quotes, 2);
        QuoteJava quote1 = CollectionUtilsJava.getFromEnd(screen_2_Quotes, 1);

        boolean lastQuotesBelowEMA = isQuoteBelowEMA(quote2, ema13_2.getValue()) && isQuoteBelowEMA(quote1, ema13_1.getValue());
        if (!lastQuotesBelowEMA) {
            LogJava.recordCode(QUOTES_NOT_BELOW_EMA_SCREEN_2, screen);
            LogJava.addDebugLine("The last two quotes are not below EMA13");
            return new BlockResultJava(screen.getLastQuote(), QUOTES_NOT_BELOW_EMA_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/ema/Long_ScreenTwo_EMA_TwoBarsBelow.kt
package com.kishlaly.ta.analyze.tasks.blocks.two.ema

import com.kishlaly.ta.analyze.tasks.blocks.two.ScreenTwoBlock
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.below
import com.kishlaly.ta.model.indicators.EMA
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.utils.CollectionUtils
import com.kishlaly.ta.utils.Log

/**
 * the last two bars are completely below EMA13
 */
class Long_ScreenTwo_EMA_TwoBarsBelow : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screen_2_EMA13 = screen.indicator(Indicator.EMA13) as List<EMA>
        val ema2 = CollectionUtils.getFromEnd<EMA>(screen_2_EMA13, 2)
        val ema1 = screen_2_EMA13.last()

        val lastQuotesBelowEMA = screen.preLastQuote below ema2 && screen.lastQuote below ema1
        if (!lastQuotesBelowEMA) {
            Log.recordCode(BlockResultCode.QUOTES_NOT_BELOW_EMA_SCREEN_2, screen)
            Log.addDebugLine("The last two quotes are not below EMA13")
            return BlockResult(screen.lastQuote, BlockResultCode.QUOTES_NOT_BELOW_EMA_SCREEN_2)
        }
        return BlockResult(screen.lastQuote, BlockResultCode.OK)

    }
}